draft-ietf-quic-transport.md

```diff
@@ -3007,9 +3007,6 @@ guidance offered below seeks to strike this balance.
 
 ### Sending ACK Frames {#sending-acknowledgements}
 
-An ACK frame SHOULD be generated for at least every second ack-eliciting packet.
-This recommendation is in keeping with standard practice for TCP {{?RFC5681}}.
-
 An endpoint MUST NOT excessively delay acknowledgements of ack-eliciting
```

Comment:
I think that we're still missing the "An endpoint MUST send at least one acknowledgment for every packet it receives and successfully processes."  This paragraph is mostly about max ack delay, which is fine, but it only really says that you can't excessively delay.

Sure, you could say that "never send" == "excessive delay", but you could also say that you can't delay something that never existed in the first place.

Reply:
What if you run out if space in the ACK frame?

Reply:
If you send an ACK for every other received packet and drop all older acknowledgments, then you use at most 19 bytes for an ACK, probably less.  I don't think that will break any limits.

Reply:
@mikkelfj Even if you only ACK the last 2 packets per ACK, QUIC should work correctly.  It's if packets never get acknowledged, especially consistently, there are potential problems.

Reply:
I can see this works on average - but if you opt to max out on max delay to minimize transmissions in one direction you could find you self in trouble. A MUST would then force to send more frequently than optimal whereas SHOULD would allow for occasionally dropping the ball. Dropping the ball frequently would be be bad, of course, suggesting that ACKs are not sent often enough.

Reply:
@mikkelfj I don't understand the issue you're concerned about.  One can fit an ACK of any two packets into a datagram, which is the minimum requirement of this text.  Yes, you could set a very large max_ack_delay, which would cause enormously long PTO timeouts and be very bad for performance in most cases.

Reply:
I don't want to make this a hard argument, I'm just slightly concerned. Maybe it is a non-issue, or maybe potential issues are shadowed by other concerns such as PTO. Some devices might not want to fire up a transmitter radio more often than necessary. But generally I'm just concerned with a MUST that cannot be enforced or fully controlled - you can easily claim it works under usual conditions but it might not under abnormal conditions. A SHOULD would make it possible to drop packets under those a conditions. A MUST would either have to make assumptions that are hard to prove correct, or have complex code to handle edge case where ACK's queue op, or violate the MUST. I suspect most implementations would go for the latter and hope they don't violate by sending ACK's often enough.